internal/attestation/aws/snp/validator.go

```diff
@@ -7,45 +7,63 @@ SPDX-License-Identifier: AGPL-3.0-only
 package snp
 
 import (
+	"bytes"
 	"context"
 	"crypto"
+	"crypto/x509"
 	"encoding/json"
+	"encoding/pem"
+	"errors"
 	"fmt"
+	"io"
+	"net/http"
 
-	awsConfig "github.com/aws/aws-sdk-go-v2/config"
-	"github.com/aws/aws-sdk-go-v2/feature/ec2/imds"
-	"github.com/aws/aws-sdk-go-v2/service/ec2"
 	"github.com/edgelesssys/constellation/v2/internal/attestation"
 	"github.com/edgelesssys/constellation/v2/internal/attestation/variant"
 	"github.com/edgelesssys/constellation/v2/internal/attestation/vtpm"
 	"github.com/edgelesssys/constellation/v2/internal/config"
+	"github.com/google/go-sev-guest/abi"
+	"github.com/google/go-sev-guest/verify"
 	"github.com/google/go-tpm-tools/proto/attest"
 	"github.com/google/go-tpm/tpm2"
 )
 
 // Validator for AWS TPM attestation.
 type Validator struct {
+	// Embed variant to identify the Validator using varaint.OID().
 	variant.AWSSEVSNP
+	// Embed validator to implement Validate method for aTLS handshake.
 	*vtpm.Validator
-	getDescribeClient func(context.Context, string) (awsMetadataAPI, error)
+	// AMD root key. Root of trust for the ASK used during report validation.
+	ark *x509.Certificate
+	// kdsClient gets an ASK from somewhere. kdsClient is required for testing.
+	kdsClient askGetter
+	// reportValidator validates a SNP report. reportValidator is required for testing.
+	reportValidator snpReportValidator
 }
 
 // NewValidator create a new Validator structure and returns it.
 func NewValidator(cfg *config.AWSSEVSNP, log attestation.Logger) *Validator {
-	v := &Validator{}
+	v := &Validator{
+		ark:       (*x509.Certificate)(&cfg.AMDRootKey),
+		kdsClient: kdsClient{http.DefaultClient},
+	}
+
 	v.Validator = vtpm.NewValidator(
 		cfg.Measurements,
-		getTrustedKey,
-		v.tpmEnabled,
+		v.getTrustedKey,
+		func(vtpm.AttestationDocument, *attest.MachineState) error { return nil },
 		log,
 	)
-	v.getDescribeClient = getEC2Client
 	return v
 }
 
 // getTrustedKeys return the public area of the provides attestation key.
-// Normally, here the trust of this key should be verified, but currently AWS does not provide this feature.
-func getTrustedKey(_ context.Context, attDoc vtpm.AttestationDocument, _ []byte) (crypto.PublicKey, error) {
+// Normally, the key should be verified here, but currently AWS does not provide means to do so.
+func (v *Validator) getTrustedKey(ctx context.Context, attDoc vtpm.AttestationDocument, userData []byte) (crypto.PublicKey, error) {
+	if err := v.reportValidator.validate(ctx, attDoc, v.kdsClient, v.ark, userData); err != nil {
```

Comment:
I think it makes more sense to put the TPM pubkey in the SNP report user data so we have a chain
SNP report -> verifies -> TPM -> verifies -> actual userData
similar how it works on Azure. Do you agree @derpsteb @daniel-weisse ? Or am I missing something?

Reply:
Technically there shouldn't be a difference in security (yet), but I agree that the story makes a bit more sense that way.

Reply:
I agree with Daniel's conclusion. Nevertheless, I changed the code to include the sha512 hash of the AK (pub) in the SNP report. The validator calculates the hash of the AK it received in the attestation statement and compares it to the userdata.